community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/ast/convert/plannerQuery/ClauseConverters.scala

```diff
@@ -381,6 +384,28 @@ object ClauseConverters {
       ).
       withTail(PlannerQuery.empty)
 
+  private def addForeachToLogicalPlanInput(builder: PlannerQueryBuilder, clause: Foreach): PlannerQueryBuilder = {
+    val setOfVariableNode =
+      if (builder.semanticTable.isNode(clause.variable))
+        Set(IdName.fromVariable(clause.variable))
+      else Set.empty
+
+    val innerBuilder = new PlannerQueryBuilder(PlannerQuery.empty, builder.semanticTable)
+      .amendQueryGraph(_.addPatternNodes((builder.allSeenPatternNodes ++ setOfVariableNode).toSeq:_*)
+                        .addArgumentId(IdName.fromVariable(clause.variable)))
+
+    val innerPlannerQuery = clause.updates.foldLeft(innerBuilder) {
+      case (acc, innerClause) => addToLogicalPlanInput(acc, innerClause)
+    }.build()
+
+    val foreachPattern = ForeachPattern(
+      variable = IdName(clause.variable.name),
+      expression = clause.expression,
+      innerUpdates = innerPlannerQuery)
+
+    builder.amendQueryGraph(_.addMutatingPatterns(foreachPattern))
```

Comment:
Not in itself, but its purpose is definitely tied to updates, since it allows only update clauses in its body.